tests/validate-fs.sh

```diff
@@ -0,0 +1,42 @@
+#!/bin/bash
+
+status=0
+
+# $1 = directory
+# $2, $3, ... = permitted file extensions
+function checkExt()
+{
+  dir="$1"
+  shift
+  regex="\\.\($(echo $* | sed 's/ /\\|/g')\)$"
+  if find "$dir" -type f | grep -q -v "$regex"; then
+    echo "Directory '$dir' may only have files with the following extensions: $*"
+    status=1
+  fi
+}
+
+# $1 = directory
+# $2, $3, ... = permitted file permissions
+function checkPerm()
+{
+  dir="$1"
+  shift
+  violation=$(find $dir -type f -exec stat -c '%a' '{}' +| grep -vq "$(echo $* | sed 's/ /|/g')")
+  for file in $violation; do
+    echo "Directory '$dir' may only have files with the following permissions: $*"
+    status=1
+  done
```

Comment:
```suggestion
  allowed="$@"
  unset pattern
  for p; do
    pattern+="${pattern+ -a }! -perm $p"
  done
  if find $dir -type f $pattern | grep -q .; then
    echo "Directory '$dir' may only have files with the following permissions: $allowed"
    status=1
  fi
```
The `-perm` option for `find` is much faster. Also, the error message should only be displayed once per directory, not once for each offending file.

Reply:
Thanks! Great to have a shell wizard on the team :smile: